pymc/gp/hsgp.py

```diff
@@ -0,0 +1,357 @@
+            X = np.linspace(0, 10, 100)[:, None]
+
+            with pm.Model() as model:
+                eta = pm.Exponential("eta", lam=1.0)
+                ell = pm.InverseGamma("ell", mu=5.0, sigma=5.0)
+                cov_func = eta**2 * pm.gp.cov.ExpQuad(1, ls=ell)
+
+                # m = [200] means 200 basis vectors for the first dimenison
+                # L = [10] means the approximation is valid from Xs = [-10, 10]
+                gp = pm.gp.HSGP(m=[200], L=[10], cov_func=cov_func)
+
+                # Order is important.  First calculate the mean, then make X a shared variable,
+                # then subtract the mean.  When X is mutated later, the correct mean will be
+                # subtracted.
+                X_mean = np.mean(X, axis=0)
+                X = pm.MutableData("X", X)
+                Xs = X - X_mean
+
+                # Pass the zero-subtracted Xs in to the GP
+                phi, sqrt_psd = gp.prior_linearized(Xs=Xs)
+
+                # Specify standard normal prior in the coefficients.  The number of which
+                # is given by the number of basis vectors, which is also saved in the GP object
+                # as m_star.
+                beta = pm.Normal("beta", size=gp.m_star)
+
+                # The (non-centered) GP approximation is given by
+                f = pm.Deterministic("f", phi @ (beta * sqrt_psd))
+
+                ...
+
+
+            # Then it works just like a linear regression to predict on new data.
+            # First mutate the data X,
+            x_new = np.linspace(-10, 10, 100)
+            with model:
+                model.set_data("X", x_new[:, None])
+
+            # and then make predictions for the GP using posterior predictive sampling.
+            with model:
+                ppc = pm.sample_posterior_predictive(idata, var_names=["f"])
+        """
+
+        # Index Xs using input_dim and active_dims of covariance function
+        Xs, _ = self.cov_func._slice(Xs)
+
+        # If not provided, use Xs and c to set L
+        if self._L is None:
+            assert isinstance(self._c, (float, np.ndarray, pt.TensorVariable))
+            self.L = set_boundary(Xs, self._c)
+        else:
+            self.L = self._L
+
+        eigvals = calc_eigenvalues(self.L, self._m, tl=pt)
+        phi = calc_eigenvectors(Xs, self.L, eigvals, self._m, tl=pt)
+        omega = pt.sqrt(eigvals)
+        psd = self.cov_func.power_spectral_density(omega)
+
+        i = int(self._drop_first == True)
+        return phi[:, i:], pt.sqrt(psd[i:])
+
+    def prior(self, name: str, X: TensorVariable, *args, **kwargs):
+        R"""
+        Returns the (approximate) GP prior distribution evaluated over the input locations `X`.
+
+        Parameters
+        ----------
+        name: string
+            Name of the random variable
+        X: array-like
+            Function input values.
+        dims: None
+            Dimension name for the GP random variable.
+        """
```

Comment:
See similar comments below:
* args/kwargs getting dropped silently
* Parameters section formatting
* Docstring doesn't match signature

Reply:
Right, I added the args/kwargs issues you're pointing to make mypy pass.  It looks like I then forgot to update the docstrings.  It originally just passed dims like the docstring says.  Without doing this mypy complains that the HSGP function signatures don't match the signatures of the base class.  Is there a third option?  I could have HSGP not be a subclass the base gp class, which seems weird to me because it is a subclass. 

Is it actually a bad thing if args/kwargs get dropped silently?  At least with kwargs isnt that expected?  The docstring says it just takes dims, so what else should a user expect?

Reply:
I'd expect it to take any of the kwargs of a typical PyMC distribution, for example `dims`, `initval`...

Can you make the signature of the base class more specific? (Specifying just `dims`, not `*args, **kwargs`..)

Otherwise you can go with a `# type: ignore` comment, but be aware that this violates the Liskov subtitution principle. Best check https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides

Reply:
Great, added the ignores.  Maybe in a later PR I can refactor the GP module to not violate the liskov substitution principle.

Reply:
args/kwargs arent getting dropped silently now, and the docstrings match the signature.  I also added a `Returns` text to the docstring for `prior_linearized`.  

I held off on adding `Returns` to the docstrings for the rest of the methods because the rest of the GP module methods don't have a returns sections either.  I think it's good to have of course, but out of scope to add it for everything here.  I also would like to defer refactoring how `Base` is used so it doesn't violate the Liskov sub. principle you pointed out because that also touches the entire gp submodule.  It would be good to fix, but it's been that way for several years now without issue because users don't use `Base`.  But I agree, structurally it could be better.